bounded-collections/src/bounded_btree_map.rs

```diff
@@ -113,6 +113,7 @@ where
 			return Err("BoundedBTreeMap exceeds its limit".into());
 		}
 		input.descend_ref()?;
+		input.on_before_alloc_mem(codec::mem_size_of_btree::<(K, V)>(len))?;
 		let inner = Result::from_iter((0..len).map(|_| Decode::decode(input)))?;
 		input.ascend_ref();
```

Comment:
Should be replaced by

```
struct PrependCompactInput { compact: Compact<u32>, read: u32, inner: I }

/// impl Input that first prepends the compact and then uses the inner

/// After that you use it

let inner = BTreeMap::decode(&mut PrependCompactInput { compact, read: 0, inner: input })?
```

Reply:
Nice idea, implemented as you've described it in [1a73132](https://github.com/paritytech/parity-common/pull/906/commits/1a7313228a8b930a28ac85e440efee22a69a21eb)

Depending on how hard we want to go we could maybe avoid some allocations since `Compact<u32>` has a fixed upper limit on its size with something like:

```
struct PrependCompactInput<'a, I> {
    compact_bytes: [u8; 5],
    compact_len: u8,
    read: u32,
    inner: &'a mut I,
}
```
Then you just encode the compact once instead of on each read. When we construct the struct we dump the compact into an array and hardcode its length, so we know only the first `compact_len` bytes are valid